### PR TITLE
Update checks to be consistent with multi-batch implementation

### PR DESCRIPTION
**Pull request by fantaosha**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: theseus/labs/lie/functional/se3_impl.py [hunk 1]
```diff
@@ -21,85 +21,68 @@
...
     def _impl(t_: torch.Tensor):
-        if t_.ndim != 3 or t_.shape[1:] != (3, 4):
-            raise ValueError(
-                f"SE3 data tensors can only be 3x4 matrices, but got shape {t_.shape}."
-            )
         SO3.check_group_tensor(t_[:, :, :3])
 
+    if tensor.shape[-2:] != (3, 4):
+        raise ValueError(
+            f"SE3 data tensors can only be 3x4 matrices, but got shape {tensor.shape}."
```
Comment (luisenp) [inline theseus/labs/lie/functional/se3_impl.py:28]: Now that we support multiple dimensions, I think a better message for all of these will be
```python
f"SE3 data tensors can must have shape (..., 3, 4) but got shape {tensor.shape}."
```

File: theseus/labs/lie/functional/se3_impl.py [hunk 2]
```diff
@@ -21,85 +21,68 @@
...
+        )
 
 
 def check_hat_matrix(matrix: torch.Tensor):
     def _impl(t_: torch.Tensor):
-        if t_.ndim != 3 or t_.shape[1:] != (4, 4):
-            raise ValueError("Hat matrices of SE(3) can only be 3x4 matrices")
-
-        if t_[:, -1].abs().max() > constants._SE3_NEAR_ZERO_EPS[t_.dtype]:
+        if matrix[:, -1].abs().max() > constants._SE3_NEAR_ZERO_EPS[matrix.dtype]:
```
Comment (luisenp) [inline theseus/labs/lie/functional/se3_impl.py:57]: Shouldn't this have `matrix` -> `t_` and `:` -> `...`?

File: theseus/labs/lie/functional/se3_impl.py [hunk 3]
```diff
@@ -21,85 +21,68 @@
...
             raise ValueError("The last row for hat matrices of SE(3) must be zero")
 
         SO3.check_hat_matrix(t_[:, :3, :3])
```
Comment (luisenp) [inline theseus/labs/lie/functional/se3_impl.py:60]: Missing ellipsis here?

File: theseus/labs/lie/functional/so3_impl.py [hunk 4]
```diff
@@ -66,47 +60,41 @@ def check_transform_tensor(tensor: torch.Tensor):
 
 def check_hat_matrix(matrix: torch.Tensor):
     def _impl(t_: torch.Tensor):
-        if t_.ndim != 3 or t_.shape[1:] != (3, 3):
-            raise ValueError("Hat matrices of SO(3) can only be 3x3 matrices")
-
         if (t_.transpose(1, 2) + t_).abs().max().item() > constants._SO3_HAT_EPS[
```
Comment (luisenp) [inline theseus/labs/lie/functional/so3_impl.py:63]: `transpose(-1, -2)`?

**Review by luisenp — CHANGES_REQUESTED** [review]
Looks mostly good. Some small comments. Thanks!

File: theseus/labs/lie/functional/so3_impl.py (hunk 4 above)
Comment (fantaosha) [reply theseus/labs/lie/functional/so3_impl.py]: Fixed, thanks!

File: theseus/labs/lie/functional/se3_impl.py (hunk 2 above)
Comment (fantaosha) [reply theseus/labs/lie/functional/se3_impl.py]: Ah, good catch.

File: theseus/labs/lie/functional/se3_impl.py (hunk 3 above)
Comment (luisenp) [reply theseus/labs/lie/functional/se3_impl.py]: What do you mean? Shouldn't it be `check_hat_matrix(t_[..., :3, :3])`

**Review by luisenp — CHANGES_REQUESTED** [review]
Only one comment, thanks!

File: theseus/labs/lie/functional/se3_impl.py (hunk 3 above)
Comment (fantaosha) [reply theseus/labs/lie/functional/se3_impl.py]: > What do you mean? Shouldn't it be `check_hat_matrix(t_[..., :3, :3])`

Ah, sorry. My fault. It has been fixed.

File: theseus/labs/lie/functional/se3_impl.py [hunk 5]
```diff
@@ -21,85 +21,68 @@
...
-                f"Tangent vectors of SE3 should be 6-D vectors, "
-                f"but got shape {t_.shape}."
-            )
-
-    checks_base(tangent_vector, _impl)
+    _check = tangent_vector.shape[-2:] == (6, 1)
+    _check |= tangent_vector.shape[-1] == 6
+    if not _check:
+        raise ValueError(
+            f"Tangent vectors of SE3 should be 6-D vectors, "
```
[the remark this answers is not in the document]
Comment (fantaosha) [reply theseus/labs/lie/functional/se3_impl.py]: Updated all the error messages in the same format.

File: theseus/labs/lie/functional/se3_impl.py [hunk 6]
```diff
@@ -21,85 +21,71 @@
...
-        if t_.ndim != 3 or t_.shape[1:] != (3, 4):
-            raise ValueError(
-                f"SE3 data tensors can only be 3x4 matrices, but got shape {t_.shape}."
-            )
-        SO3.check_group_tensor(t_[:, :, :3])
+        SO3.check_group_tensor(t_[..., :3])
+
+    if tensor.shape[-2:] != (3, 4):
+        raise ValueError(
+            f"SE3 data tensors can must have shape (..., 3, 4) but got shape {tensor.shape}."
```
Comment (luisenp) [inline theseus/labs/lie/functional/se3_impl.py:28]: "can must" -> "must"

File: theseus/labs/lie/functional/se3_impl.py [hunk 7]
```diff
@@ -21,85 +21,71 @@
...
-    def _impl(t_):
-        if t_.ndim != 3 or t_.shape[-2:] != (3, 4):
-            raise ValueError(
-                f"SE3 data tensors can only be 3x4 matrices, but got shape {t_.shape}."
-            )
-
-    checks_base(tensor, _impl)
+    if tensor.shape[-2:] != (3, 4):
+        raise ValueError(
+            f"SE3 data tensors can must have shape (..., 3, 4) but got shape {tensor.shape}."
```
Comment (luisenp) [inline theseus/labs/lie/functional/se3_impl.py:37]: "can must" -> "must"

File: theseus/labs/lie/functional/se3_impl.py [hunk 8]
```diff
@@ -21,85 +21,71 @@
...
-                f"Tangent vectors of SE3 should be 6-D vectors, "
-                f"but got shape {t_.shape}."
-            )
-
-    checks_base(tangent_vector, _impl)
+    _check = tangent_vector.shape[-2:] == (6, 1)
+    _check |= tangent_vector.shape[-1] == 6
+    if not _check:
+        raise ValueError(
+            f"Tangent vectors of SE3 must have shape (..., 6) or (..., 6, 1), "
```
Comment (luisenp) [inline theseus/labs/lie/functional/se3_impl.py:50]: the other messages don't have "," before "but got shape". Let's remove it for consistency.

File: theseus/labs/lie/functional/se3_impl.py [hunk 9]
```diff
@@ -21,85 +21,71 @@
...
+            raise ValueError("The last row for hat matrices of SE3 must be zero")
 
-        if t_[:, -1].abs().max() > constants._SE3_NEAR_ZERO_EPS[t_.dtype]:
-            raise ValueError("The last row for hat matrices of SE(3) must be zero")
+        SO3.check_hat_matrix(t_[..., :3, :3])
 
-        SO3.check_hat_matrix(t_[:, :3, :3])
+    if matrix.shape[-2:] != (4, 4):
+        raise ValueError(
+            f"Hat matrices of SE3 must have shape (..., 4, 4), "
```
Comment (luisenp) [inline theseus/labs/lie/functional/se3_impl.py:64]: Remove "," as above.

File: theseus/labs/lie/functional/so3_impl.py [hunk 10]
```diff
@@ -21,41 +21,42 @@
...
             - torch.eye(3, 3, dtype=t_.dtype, device=t_.device)
         ).abs().max().item() < MATRIX_EPS
         _check &= (torch.linalg.det(t_) - 1).abs().max().item() < MATRIX_EPS
 
         if not _check:
             raise ValueError("Invalid data tensor for SO3.")
 
+    if tensor.shape[-2:] != (3, 3):
+        raise ValueError(
+            f"SO3 data tensors can must have shape (..., 3, 3) but got shape {tensor.shape}."
```
Comment (luisenp) [inline theseus/labs/lie/functional/so3_impl.py:39]: "can must" -> "must"

File: theseus/labs/lie/functional/so3_impl.py [hunk 11]
```diff
@@ -21,41 +21,42 @@
...
 
 def check_matrix_tensor(tensor: torch.Tensor):
-    def _impl(t_: torch.Tensor):
-        if t_.ndim != 3 or t_.shape[-2:] != (3, 3):
-            raise ValueError("Matrix tensors can only be 3x3 matrices.")
-
-    checks_base(tensor, _impl)
+    if tensor.shape[-2:] != (3, 3):
+        raise ValueError(
+            f"SO3 data tensors can must have shape (..., 3, 3) but got shape {tensor.shape}."
```
Comment (luisenp) [inline theseus/labs/lie/functional/so3_impl.py:48]: "can must" -> "must"

File: theseus/labs/lie/functional/so3_impl.py [hunk 12]
```diff
@@ -21,41 +21,42 @@
...
-        _check |= t_.ndim == 2 and t_.shape[1] == 3
-        if not _check:
-            raise ValueError("Tangent vectors of SO3 should be 3-D vectors.")
-
-    checks_base(tangent_vector, _impl)
+    _check = tangent_vector.shape[-2:] == (3, 1)
+    _check |= tangent_vector.shape[-1] == 3
+    if not _check:
+        raise ValueError(
+            f"Tangent vectors of SO3 must have shape (..., 3) or (..., 3, 1), "
```
Comment (luisenp) [inline theseus/labs/lie/functional/so3_impl.py:57]: Remove "," as above.

File: theseus/labs/lie/functional/so3_impl.py [hunk 13]
```diff
@@ -66,47 +67,48 @@ def check_transform_tensor(tensor: torch.Tensor):
...
-        if (t_.transpose(1, 2) + t_).abs().max().item() > constants._SO3_HAT_EPS[
+        if (t_.transpose(-1, -2) + t_).abs().max().item() > constants._SO3_HAT_EPS[
             t_.dtype
         ]:
-            raise ValueError("Hat matrices of SO(3) can only be skew-symmetric.")
+            raise ValueError("Hat matrices of SO3 can only be skew-symmetric.")
+
+    if matrix.shape[-2:] != (3, 3):
+        raise ValueError(
+            f"Hat matrices of SO3 must have shape (..., 3, 3), "
```
Comment (luisenp) [inline theseus/labs/lie/functional/so3_impl.py:77]: Remove "," as above.

File: theseus/labs/lie/functional/so3_impl.py [hunk 14]
```diff
@@ -66,47 +67,48 @@ def check_transform_tensor(tensor: torch.Tensor):
...
         if t_.dtype != torch.float64:
             t_ = t_.double()
 
-        if (torch.linalg.norm(t_, dim=1) - 1).abs().max().item() >= QUANTERNION_EPS:
+        if (torch.linalg.norm(t_, dim=-1) - 1).abs().max().item() >= QUANTERNION_EPS:
             raise ValueError("Not unit quaternions.")
 
+    if quaternion.shape[-1] != 4:
+        raise ValueError(
+            f"Quaternions must have shape (..., 4), "
```
Comment (luisenp) [inline theseus/labs/lie/functional/so3_impl.py:96]: Remove "," as above.

File: theseus/labs/lie/functional/so3_impl.py [hunk 15]
```diff
@@ -66,47 +67,48 @@ def check_transform_tensor(tensor: torch.Tensor):
...
+            f"but got shape {quaternion.shape}."
+        )
+        raise ValueError("Quaternions can only be 4-D vectors.")
```
Comment (luisenp) [inline theseus/labs/lie/functional/so3_impl.py:99]: Forgot to remove the old `ValueError`

**Review by luisenp — CHANGES_REQUESTED** [review]
Found a bunch of typos, probably from copy paste. I think it's better if we template this for consistency. Something like this:

```python
def _shape_err_msg(data_type: str, expected_shape: str, tensor_shape: torch.Size) -> str:
    return f"{data_type} must have shape {expected_shape} but got shape {tensor_shape}."
```

This will make it easier to add the same message to the other methods like lift, project, etc, that currently have a different style. Usage will be 

```python
raise ValueError(_shape_error_msg("Left act matrices of SO3", "(..., 3, -1)", matrix.shape))
```

**Review by luisenp — APPROVED** [review]
LGTM. Thanks for addressing all the comments.

**Pull request merged**